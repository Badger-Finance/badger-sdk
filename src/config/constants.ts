import { ARB_CONFIG } from './network/arbitrum.config';
import { AVAX_CONFIG } from './network/avalanche.config';
import { BSC_CONFIG } from './network/binance.config';
import { ETH_CONFIG } from './network/ethereum.config';
import { FTM_CONFIG } from './network/fantom.config';
import { LOCAL_CONFIG } from './network/local.config';
import { MATIC_CONFIG } from './network/polygon.config';

export const ONE_MIN_MS = 60 * 1000;
export const ONE_HOUR_MS = 60 * ONE_MIN_MS;
export const ONE_DAY_MS = 24 * ONE_HOUR_MS;
export const ONE_YEAR_MS = 365 * ONE_DAY_MS;

export const SUPPORTED_NETWORKS = [
  ETH_CONFIG,
  LOCAL_CONFIG,
  BSC_CONFIG,
  MATIC_CONFIG,
  ARB_CONFIG,
  AVAX_CONFIG,
  FTM_CONFIG,
];

export const VERSION = '1.0.77';
