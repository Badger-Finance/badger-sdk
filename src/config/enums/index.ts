export { Network } from './network.enum';
