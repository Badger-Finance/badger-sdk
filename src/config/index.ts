export * from './enums';
export * from './types';
export * from './constants';
export * from './interfaces';
export * from './network';
