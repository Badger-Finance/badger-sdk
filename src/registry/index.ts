export * from './enums';
export * from './interfaces';
export { RegistryService } from './registry.service';
