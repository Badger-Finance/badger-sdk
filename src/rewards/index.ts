export * from './interfaces';
export { RewardsService } from './rewards.service';
