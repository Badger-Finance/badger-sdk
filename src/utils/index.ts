export * from './chunk-query-filter';
export * from './deployed-at.util';
export * from './key-by';
