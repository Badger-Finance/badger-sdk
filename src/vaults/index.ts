export * from './enums';
export * from './interfaces';
export { VaultsService } from './vaults.service';
export * from './vaults.utils';
