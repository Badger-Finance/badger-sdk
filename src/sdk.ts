import { ethers } from 'ethers';
import { providers } from '@0xsequence/multicall';

import { Signer } from '@ethersproject/abstract-signer';
import { Networkish } from '@ethersproject/providers';

import { APIOptions, BadgerAPI } from './api';
import { BadgerGraph } from './graphql';

import { NetworkConfig } from './config/network/network.config';

import { DiggService } from './digg/digg.service';
import { ibBTCService } from './ibbtc/ibbtc.service';
import { RegistryService } from './registry/registry.service';
import { RewardsService } from './rewards/rewards.service';
import { VaultsService } from './vaults/vaults.service';
import { TokensService } from './tokens/tokens.service';

import { SDKProvider } from './config/types/sdk-provider';

import { SUPPORTED_NETWORKS } from './config/constants';

export interface SDKOptions extends APIOptions {
  provider: SDKProvider | string;
}

export class BadgerSDK {
  private static initialized = false;
  private loading: Promise<void>;

  public address?: string;
  public api: BadgerAPI;
  public graph: BadgerGraph;
  public config: NetworkConfig;
  public provider: providers.MulticallProvider;
  public signer?: Signer;

  readonly registry: RegistryService;
  readonly digg: DiggService;
  readonly ibbtc: ibBTCService;
  readonly rewards: RewardsService;
  readonly tokens: TokensService;
  readonly vaults: VaultsService;

  constructor({ network, provider, baseURL }: SDKOptions) {
    if (!BadgerSDK.initialized) {
      for (const config of SUPPORTED_NETWORKS) {
        NetworkConfig.register(config);
      }
      BadgerSDK.initialized = true;
    }

    const sdkProvider = BadgerSDK.getSdkProvider(provider);

    this.provider = new providers.MulticallProvider(sdkProvider);
    this.signer = sdkProvider.getSigner();
    this.loading = this.initialize();
    this.config = NetworkConfig.getConfig(network);
    this.api = new BadgerAPI({
      network: this.config.network,
      baseURL,
    });
    this.graph = new BadgerGraph({
      network: this.config.network,
    });
    this.digg = new DiggService(this);
    this.ibbtc = new ibBTCService(this);
    this.registry = new RegistryService(this);
    this.rewards = new RewardsService(this);
    this.tokens = new TokensService(this);
    this.vaults = new VaultsService(this);
  }

  ready() {
    return Promise.all([this.loading, this.registry.ready(), this.rewards.ready()]);
  }

  update(network: Networkish, provider: SDKProvider) {
    this.updateNetwork(network);
    this.updateProvider(provider);
  }

  updateNetwork(network: Networkish) {
    this.config = NetworkConfig.getConfig(network);
    this.api = new BadgerAPI({
      network: this.config.network,
      baseURL: this.api.baseURL,
    });
    this.graph = new BadgerGraph({
      network: this.config.network,
    });
  }

  updateProvider(provider: SDKProvider) {
    const sdkProvider = BadgerSDK.getSdkProvider(provider);

    this.provider = new providers.MulticallProvider(sdkProvider);
    this.signer = sdkProvider.getSigner();
  }

  private static getSdkProvider(provider: SDKProvider | string): SDKProvider {
    let sdkProvider: SDKProvider;

    if (typeof provider === 'string') {
      sdkProvider = new ethers.providers.JsonRpcBatchProvider(provider);
    } else {
      sdkProvider = provider;
    }

    return sdkProvider;
  }

  private async initialize() {
    try {
      if (this.signer) {
        this.address = await this.signer.getAddress();
      }
    } catch {} // ignore errors from getting address
  }
}
