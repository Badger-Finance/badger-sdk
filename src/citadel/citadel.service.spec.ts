import { BadgerSDK } from '../sdk';
import { mock, MockProxy } from 'jest-mock-extended';
import {
  CitadelMinter,
  CitadelMinter__factory,
  StakedCitadelLocker,
  StakedCitadelLocker__factory,
} from '../contracts';
import { JsonRpcProvider, JsonRpcSigner } from '@ethersproject/providers';
import { Network } from '../config';
import { RegistryService } from '../registry';
import { RewardsService } from '../rewards';
import * as vaultUtils from '../vaults/vaults.utils';
import { citadelDistributionToStakingEvents } from './mocks/distribution-to-staking-events.mock';
import { addedRewardEvents } from './mocks/added-reward-events.mock';
import { paidRewardEvents } from './mocks/paid-reward-events.mock';
import { RewardFilter } from './enums/reward-filter.enum';

describe('citadel.service', function () {
  const TEST_ADDR = '0x96d4dBdc91Bef716eb407e415c9987a9fAfb8906';
  const TEST_TOKEN = '0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599'; // WBTC
  const TEST_USER = '0x3472A5A71965499acd81997a54BBA8D852C6E53d'; // Badger

  const DEPLOYED_AT = 10;
  const CURRENT_BLOCK = 150;

  let sdk: BadgerSDK;
  let minter: MockProxy<CitadelMinter>;
  let locker: MockProxy<StakedCitadelLocker>;

  beforeEach(async () => {
    // Setup Jest Mocks
    const mockSigner = mock<JsonRpcSigner>();
    mockSigner.getAddress
      .calledWith()
      .mockImplementation(async () => TEST_ADDR);
    const mockProvider = mock<JsonRpcProvider>();
    mockProvider.getSigner.calledWith().mockImplementation(() => mockSigner);
    mockProvider.getBlockNumber
      .calledWith()
      .mockImplementation(async () => CURRENT_BLOCK);

    // Services that will force contracts connection in sdk constructor
    jest.spyOn(RegistryService.prototype, 'ready').mockImplementation();
    jest.spyOn(RewardsService.prototype, 'ready').mockImplementation();

    minter = mock<CitadelMinter>();
    locker = mock<StakedCitadelLocker>();

    mock(minter.filters)
      .CitadelDistributionToStaking.calledWith()
      .mockImplementation();

    mock(locker.filters).RewardAdded.calledWith().mockImplementation();
    mock(locker.filters).RewardPaid.calledWith().mockImplementation();

    jest
      .spyOn(CitadelMinter__factory, 'connect')
      .mockImplementation(() => minter);
    jest
      .spyOn(StakedCitadelLocker__factory, 'connect')
      .mockImplementation(() => locker);

    sdk = new BadgerSDK({
      network: Network.Ethereum,
      provider: mockProvider,
    });

    // Let the SDK Prepare
    await sdk.ready();

    jest
      .spyOn(vaultUtils, 'vaultBlockDeployedAt')
      .mockImplementation(() => DEPLOYED_AT);
  });

  describe('Work only with Ethereum network', () => {
    it('method should throw error', () => {
      const mockSigner = mock<JsonRpcSigner>();
      mockSigner.getAddress
        .calledWith()
        .mockImplementation(async () => TEST_ADDR);
      const mockProvider = mock<JsonRpcProvider>();
      mockProvider.getSigner.calledWith().mockImplementation(() => mockSigner);

      sdk = new BadgerSDK({
        network: Network.Fantom,
        provider: mockProvider,
      });

      expect(async () => sdk.citadel.balanceOf(TEST_ADDR)).rejects.toThrow(
        Error,
      );
    });
  });

  describe('listDistributions', () => {
    it('should return distribution events, mapped by time', async () => {
      const asyncMock = async () => citadelDistributionToStakingEvents;
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      minter.queryFilter.calledWith().mockImplementation(asyncMock);

      const distributions = await sdk.citadel.listDistributions();

      expect(Object.keys(distributions).length).toBe(3);
      expect(distributions).toMatchSnapshot();
    });

    it('should respect start block at param', async () => {
      const asyncMock = async (_: never, fromBlock: number) => {
        return citadelDistributionToStakingEvents.filter(
          (e) => e.block >= fromBlock,
        );
      };
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      minter.queryFilter.calledWith().mockImplementation(asyncMock);

      const distributions = await sdk.citadel.listDistributions({
        startBlock: 52,
      });

      expect(Object.keys(distributions).length).toBe(2);
      expect(distributions).toMatchSnapshot();
    });
  });

  describe('listRewards', () => {
    it('should return added rewards', async () => {
      const asyncMock = async () => addedRewardEvents;
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      locker.queryFilter.calledWith().mockImplementation(asyncMock);

      const rewards = await sdk.citadel.listRewards({ token: TEST_TOKEN });

      expect(rewards.length).toBe(5);
      expect(rewards).toMatchSnapshot();
    });

    it('should return paid rewards', async () => {
      const asyncMock = async () => paidRewardEvents;
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      locker.queryFilter.calledWith().mockImplementation(asyncMock);

      const rewards = await sdk.citadel.listRewards({
        token: TEST_TOKEN,
        user: TEST_USER,
        filter: RewardFilter.PAID,
      });

      expect(rewards.length).toBe(3);
      expect(rewards).toMatchSnapshot();
    });

    it('should failt without token param', async () => {
      await expect(async () => await sdk.citadel.listRewards()).rejects.toThrow(
        Error,
      );
    });

    it('should failt without user addr for paid', async () => {
      await expect(
        async () =>
          await sdk.citadel.listRewards({
            token: TEST_TOKEN,
            filter: RewardFilter.PAID,
          }),
      ).rejects.toThrow(Error);
    });
  });
});
