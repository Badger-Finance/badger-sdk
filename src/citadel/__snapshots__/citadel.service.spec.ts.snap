// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`citadel.service listDistributions should respect start block at param 1`] = `
Object {
  "1650727090814": Array [
    Object {
      "block": 52,
      "citadelAmount": Object {
        "hex": "0x087a22",
        "type": "BigNumber",
      },
      "endTime": 1650727116602,
      "startTime": 1650727090814,
    },
    Object {
      "block": 52,
      "citadelAmount": Object {
        "hex": "0x0c",
        "type": "BigNumber",
      },
      "endTime": 1650727116699,
      "startTime": 1650727090814,
    },
  ],
  "1650727090999": Array [
    Object {
      "block": 55,
      "citadelAmount": Object {
        "hex": "0x0c",
        "type": "BigNumber",
      },
      "endTime": 1650727116699,
      "startTime": 1650727090999,
    },
  ],
}
`;

exports[`citadel.service listDistributions should return distribution events, mapped by time 1`] = `
Object {
  "1650726066923": Array [
    Object {
      "block": 50,
      "citadelAmount": Object {
        "hex": "0x1388",
        "type": "BigNumber",
      },
      "endTime": 1650726976504,
      "startTime": 1650726066923,
    },
    Object {
      "block": 50,
      "citadelAmount": Object {
        "hex": "0x0159",
        "type": "BigNumber",
      },
      "endTime": 1650726976509,
      "startTime": 1650726066923,
    },
  ],
  "1650727090814": Array [
    Object {
      "block": 52,
      "citadelAmount": Object {
        "hex": "0x087a22",
        "type": "BigNumber",
      },
      "endTime": 1650727116602,
      "startTime": 1650727090814,
    },
    Object {
      "block": 52,
      "citadelAmount": Object {
        "hex": "0x0c",
        "type": "BigNumber",
      },
      "endTime": 1650727116699,
      "startTime": 1650727090814,
    },
  ],
  "1650727090999": Array [
    Object {
      "block": 55,
      "citadelAmount": Object {
        "hex": "0x0c",
        "type": "BigNumber",
      },
      "endTime": 1650727116699,
      "startTime": 1650727090999,
    },
  ],
}
`;

exports[`citadel.service listRewards should return added rewards 1`] = `
Array [
  Object {
    "block": 50,
    "reward": Object {
      "hex": "0x64",
      "type": "BigNumber",
    },
    "token": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",
  },
  Object {
    "block": 50,
    "reward": Object {
      "hex": "0xc8",
      "type": "BigNumber",
    },
    "token": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",
  },
  Object {
    "block": 51,
    "reward": Object {
      "hex": "0x012c",
      "type": "BigNumber",
    },
    "token": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",
  },
  Object {
    "block": 52,
    "reward": Object {
      "hex": "0x0190",
      "type": "BigNumber",
    },
    "token": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",
  },
  Object {
    "block": 52,
    "reward": Object {
      "hex": "0x01f4",
      "type": "BigNumber",
    },
    "token": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",
  },
]
`;

exports[`citadel.service listRewards should return paid rewards 1`] = `
Array [
  Object {
    "block": 50,
    "reward": Object {
      "hex": "0x64",
      "type": "BigNumber",
    },
    "token": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",
    "user": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
  },
  Object {
    "block": 50,
    "reward": Object {
      "hex": "0xc8",
      "type": "BigNumber",
    },
    "token": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",
    "user": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
  },
  Object {
    "block": 51,
    "reward": Object {
      "hex": "0x012c",
      "type": "BigNumber",
    },
    "token": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",
    "user": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
  },
]
`;
